packages/babel-traverse/src/path/lib/hoister.js

```diff
@@ -32,6 +32,15 @@ const referenceVisitor = {
     const binding = path.scope.getBinding(path.node.name);
     if (!binding) return;
 
+    // we can handle reassignments only if they happen in the same scope as the declaration
+    for (const violation of binding.constantViolations) {
+      if (violation.scope !== binding.path.scope) {
```

Comment:
Can't we still hoist to the level of the violation?

```js
let foo = 'hello';
export const Component = () => {
  foo = 'goodbye';
  return <span>{array.map(() => <inner>{foo}</inner>}</span>;
};

Reply:
No, we can't. Consider this example:

```js
let foo = 0;
export const Component = () => {
  foo++;
  return Promise.resolve().then(() => <span>{foo}</span>);
};

Component();
Component();
```

Here both the components should use `2`, but if we hoist it the first one will use `1`.